src/transformStateWithClones.js

```diff
@@ -8,6 +8,38 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const act of actions) {
+    switch (act.type) {
+      case 'addProperties':
+        currentState = {
+          ...currentState,
+          ...act.extraData,
+        };
+        break;
+
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (const key of act.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+
+      case 'clear':
+        currentState = {};
+        break;
+
+      default:
+        break;
+    }
+
+    stateHistory.push(currentState);
```

Comment:
Currently, you are pushing the `currentState` object directly into the `stateHistory` array. This can lead to unintended mutations because objects are pushed by reference. To avoid this, you should push a copy of `currentState` instead, like this: `stateHistory.push({ ...currentState });`.